source/Missions/Services/Agents/AgentPublisher.cs

```diff
@@ -0,0 +1,134 @@
+﻿using Common.Logging;
+using Common.Messaging;
+using HarmonyLib;
+using Missions.Services.Agents.Messages;
+using Missions.Services.Agents.Packets;
+using Missions.Services.Network;
+using Serilog;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using TaleWorlds.MountAndBlade;
+
+namespace Missions.Services.Agents
+{
+    /// <summary>
+    /// Manages player controlled <see cref="AgentMovement"/>.
+    /// </summary>
+    public class AgentPublisher
+    {
+        private Guid _agentId;
+
+        private AgentMovement _agentMovement;
+
+        private static readonly ILogger _logger = LogManager.GetLogger<AgentPublisher>();
+
+        private IMessageBroker _messageBroker;
+
+        private readonly CancellationTokenSource _agentPollingTaskCancellationTokenSource = new CancellationTokenSource();
+
+        private readonly Task _agentPollingTask;
+
+        private readonly int _packetUpdateRate;
+
+        private readonly INetworkAgentRegistry _networkAgentRegistry;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public AgentPublisher(IMessageBroker messageBroker, int packetUpdateRate, INetworkAgentRegistry networkAgentRegistry)
+        {
+            _agentMovement = new AgentMovement(_agentId);
+            _messageBroker = messageBroker;
+
+            _agentPollingTask = Task.Run(PollAndUpdateAgentMovement, _agentPollingTaskCancellationTokenSource.Token);
+            _packetUpdateRate = packetUpdateRate;
+
+            _networkAgentRegistry = networkAgentRegistry;
+        }
+
+        ~AgentPublisher()
+        {
+            _agentPollingTaskCancellationTokenSource.Cancel();
+            _agentPollingTask.Wait();
+        }
+
+        private async Task PollAndUpdateAgentMovement()
+        {
+            while (!_agentPollingTaskCancellationTokenSource.IsCancellationRequested && Mission.Current != null)
+            {
+                // TODO: also add all player agents
+                var agents = _networkAgentRegistry.ControlledAgents.Values.ToList();
```

Comment:
Player agent will be in the controlled agents, also this does not need to be casted to a list